src/utils/analytics.utils.ts

```diff
@@ -0,0 +1,117 @@
+import { EntityManager } from 'typeorm';
+import {
+  addMonths,
+  addDays,
+  startOfYear,
+  subYears,
+  format,
+  addYears,
+} from 'date-fns';
+import { EventStatus, Frequency, Months } from 'src/common/constants';
+import { User } from 'src/github/entities/user.entity';
+import { FilterDto } from 'src/analytics/dto/filter.dto';
+
+export const buildQueryBuilder = (
+  entityManager: EntityManager,
+  userId: string,
+  filter: FilterDto,
+) => {
+  const queryBuilder = entityManager
+    .getRepository(User)
+    .createQueryBuilder('user')
+    .leftJoinAndSelect('user.repo', 'repo')
+    .leftJoinAndSelect('repo.pullRequests', 'pullRequest')
+    .where('user.githubUserId = :githubUserId', { githubUserId: userId });
+
+  if (filter?.repoName) {
+    queryBuilder.andWhere('repo.name = :name', { name: filter.repoName });
+  }
+
+  return queryBuilder;
+};
+
+export const setDateFilters = (filter: FilterDto, frequency: Frequency) => {
+  let startDate: string;
+  let endDate: string;
+  const now = new Date();
+
+  if (filter?.startDate && filter?.endDate) {
+    const start = new Date(filter.startDate);
+    const end = new Date(filter.endDate);
+    start.setHours(0, 0, 0, 0);
+    end.setHours(23, 59, 59, 999);
+    startDate = start.toISOString();
+    endDate = end.toISOString();
+  } else {
+    let start;
+    if (frequency === Frequency.MONTH) {
+      start = addMonths(now, -2);
+    } else if (frequency === Frequency.YEAR) {
+      start = startOfYear(subYears(now, 2));
+    } else {
+      start = addDays(now, -6);
+    }
+    start.setHours(0, 0, 0, 0);
+    now.setHours(23, 59, 59, 999);
+    startDate = start.toISOString();
+    endDate = now.toISOString();
+  }
+
+  return { startDate, endDate };
+};
```

Comment:
PullProofAI

The `setDateFilters` function effectively sets date filters based on the provided frequency. It would be beneficial to validate the `frequency` parameter to ensure it's a valid value.
```diff
+   if (!filter || !frequency) {
+       throw new Error('Invalid filter or frequency provided for setting date filters');
+   }
```



<!-- This is an auto-generated comment by PullProofAI -->